archive/2025-07-26-how-we-rooted-copilot-eye-research.md

```diff
@@ -0,0 +1,245 @@
+
+![Output of the command /sbin/ifconfig executed in a Python interpreter, showing network interface details like IP address, netmask, and packet statistics.](https://i0.wp.com/research.eye.security/wp-content/uploads/image-53.png?resize=667%2C412&ssl=1)
+
+We can see that it is using an OverlayFS filesystem that originates in some /legion file system path on the container host system.
+
+![Output of the command mount executed in the Python interpreter showing overlay filesystem details and mount options.](https://i0.wp.com/research.eye.security/wp-content/uploads/image-55.png?resize=740%2C595&ssl=1)
+
+Most custom scripts are located in the `/app` directory.
+
+![Output of the command 'ls -al /app' executed in the Python interpreter, showing file details such as permissions, owner, group, size, and modification date.](https://i0.wp.com/research.eye.security/wp-content/uploads/image-50.png?resize=744%2C475&ssl=1)
+
+After executing a lot of commands, it starts to really be in the mood to help out. It is then even possible to simply ask to download files or tar entire folders and provide them for download. Definitely don’t try this as the first command in a fresh session. It will throw a tantrum!
+
+It does still somehow feel the need to corrupt any binary data. Another passive-aggressive trait? Text files or command output work fine though, so base64 encoding binary files before downloading them works. Most of the time.
+
+It copies files to a working directory at `/mnt/data`, from where they can be picked up outside of the sandbox with a link that looks like `blob:https://outlook.office.com/<guid>`.
+
+![Screenshot showing a chat interface with a message from Copilot stating that the user can download the file 'generateLicense.py' and including a link for the download.](https://i0.wp.com/research.eye.security/wp-content/uploads/image-54.png?resize=567%2C197&ssl=1)
+
+The `goclientapp` binary in `/app` is a Go binary that is the interface into this container. It opens a web server on port 6000 that listens for POST requests to the `/execute` endpoint. These POST requests come in as a very simple JSON, for example after asking it to execute the `%env` command, this JSON came in.
+
+```
+{"code":"%env"}
+```
+
+The `goclientapp` binary then executes the code in the Jupyter notebook environment and we get the output. There is also a `httpproxy` go binary, that seems to function as an HTTP proxy for outgoing web traffic originating from the Jupyter environment. The container did not enable any outgoing traffic (and `ENABLE_EGRESS` was false), so this binary was probably there for a future update which would potentially allow outgoing HTTP traffic.
+
+An interesting script is `entrypoint.sh` in the `/app` directory. This seems to be the script that is executed as the entrypoint into the container, so this is running as root.
```

Comment:
This is a common issue with containerized environments. I used a similar issue to escape Zapier's code execution sandbox a few years ago ago [ZAPESCAPE](https://zenity.io/blog/research/zapescape-vulnerability-disclosure)